crates/ruff_dev/src/generate_options.rs

```diff
@@ -37,38 +96,18 @@ fn emit_field(output: &mut String, name: &str, field: &OptionField, group_name:
     output.push('\n');
 }
 
-pub(crate) fn generate() -> String {
-    let mut output: String = "### Top-level\n\n".into();
-
-    let sorted_options: Vec<_> = Options::metadata()
-        .into_iter()
-        .sorted_by_key(|(name, _)| *name)
-        .collect();
-
-    // Generate all the top-level fields.
-    for (name, entry) in &sorted_options {
-        let OptionEntry::Field(field) = entry else {
-            continue;
-        };
-        emit_field(&mut output, name, field, None);
-        output.push_str("---\n\n");
-    }
+#[derive(Default)]
+struct CollectOptionsVisitor {
+    groups: Vec<(String, OptionSet)>,
+    fields: Vec<(String, OptionField)>,
+}
 
-    // Generate all the sub-groups.
-    for (group_name, entry) in &sorted_options {
-        let OptionEntry::Group(fields) = entry else {
-            continue;
-        };
-        output.push_str(&format!("### {group_name}\n"));
-        output.push('\n');
-        for (name, entry) in fields.iter().sorted_by_key(|(name, _)| name) {
-            let OptionEntry::Field(field) = entry else {
-                continue;
-            };
-            emit_field(&mut output, name, field, Some(group_name));
-            output.push_str("---\n\n");
-        }
+impl Visit for CollectOptionsVisitor {
+    fn record_set(&mut self, name: &str, group: OptionSet) {
+        self.groups.push((name.to_owned(), group));
     }
```

Comment:
The old implementation only supports one level of nesting. This would have become a problem with `lint.pydocstyle.option` where we have two level nesting.